Kernel/Arch/x86_64/ISABus/I8042Controller.cpp

```diff
@@ -255,91 +330,106 @@ ErrorOr<void> I8042Controller::drain_output_buffer()
 
     // FIXME: Is this the correct errno value for this?
     if (self_test_result != I8042Response::Success)
         return Error::from_errno(EIO);
     return {};
 }
 
-ErrorOr<u8> I8042Controller::do_send_command(PS2Device::Type device, u8 command)
+ErrorOr<void> I8042Controller::do_send_command(PS2PortIndex port_index, u8 command)
 {
-    VERIFY(device != PS2Device::Type::Unknown);
     VERIFY(m_lock.is_locked());
 
     VERIFY(!Processor::current_in_irq());
-
-    return do_write_to_device(device, command);
+    TRY(do_write_to_device(port_index, command));
+    return {};
 }
 
-ErrorOr<u8> I8042Controller::do_send_command(PS2Device::Type device, u8 command, u8 data)
+ErrorOr<void> I8042Controller::do_send_command(PS2PortIndex port_index, u8 command, u8 data)
 {
-    VERIFY(device != PS2Device::Type::Unknown);
     VERIFY(m_lock.is_locked());
 
     VERIFY(!Processor::current_in_irq());
 
-    u8 response = TRY(do_write_to_device(device, command));
-    if (response == I8042Response::Acknowledge)
-        response = TRY(do_write_to_device(device, data));
-    return response;
+    TRY(do_write_to_device(port_index, command));
+    TRY(do_write_to_device(port_index, data));
+    return {};
 }
 
-ErrorOr<u8> I8042Controller::do_write_to_device(PS2Device::Type device, u8 data)
+ErrorOr<void> I8042Controller::do_write_to_device(PS2PortIndex port_index, u8 data)
 {
-    VERIFY(device != PS2Device::Type::Unknown);
     VERIFY(m_lock.is_locked());
 
     VERIFY(!Processor::current_in_irq());
 
     int attempts = 0;
     u8 response;
     do {
-        if (device != PS2Device::Type::Keyboard) {
+        if (port_index != I8042PS2PortIndex::FirstPort) {
+            VERIFY(port_index == I8042PS2PortIndex::SecondPort);
             TRY(prepare_for_output());
             IO::out8(I8042Port::Command, I8042Command::WriteSecondPS2PortInputBuffer);
         }
         TRY(prepare_for_output());
         IO::out8(I8042Port::Buffer, data);
 
-        response = TRY(do_wait_then_read(I8042Port::Buffer));
+        response = TRY(do_wait_then_read_any_input(I8042Port::Buffer));
     } while (response == I8042Response::Resend && ++attempts < 250);
-    if (attempts >= 250)
-        dbgln("Failed to write byte to device, gave up");
-    return response;
+    if (attempts >= 250 || response == I8042Response::Resend) {
+        dbgln("I8042: Failed to write byte to device, gave up");
+        return Error::from_errno(EBUSY);
+    }
+    return {};
 }
 
-ErrorOr<u8> I8042Controller::do_read_from_device(PS2Device::Type device)
+ErrorOr<u8> I8042Controller::do_read_from_device(PS2PortIndex port_index)
 {
-    VERIFY(device != PS2Device::Type::Unknown);
-
-    TRY(prepare_for_input(device));
+    TRY(prepare_for_input(port_index));
     return IO::in8(I8042Port::Buffer);
 }
 
-ErrorOr<void> I8042Controller::prepare_for_input(PS2Device::Type device)
+ErrorOr<void> I8042Controller::prepare_for_any_input()
```

Comment:
To make it consistent with how `prepare_for_any_input` is implemented, should you also change the name from `prepare_for_output` to `prepare_for_any_output` as it doesn't take any function parameter?

Reply:
Seems like a valid suggestion :)